File: solutions/golang/musicstreamingservice/music_streaming_service_demo.go
package musicstreamingservice

import (
	"fmt"
)

func Run() {
	service := NewMusicStreamingService()

	// Create users
	user1 := NewUser("1", "john@example.com", "password123")
	user2 := NewUser("2", "jane@example.com", "password456")

	// Create songs
	song1 := &Song{ID: "1", Title: "Song 1", Artist: "Artist 1", Album: "Album 1", Duration: 180}
	song2 := &Song{ID: "2", Title: "Song 2", Artist: "Artist 2", Album: "Album 2", Duration: 200}
	song3 := &Song{ID: "3", Title: "Song 3", Artist: "Artist 3", Album: "Album 3", Duration: 210}

	// Create albums
	album1 := &Album{ID: "1", Title: "Album 1", Artist: "Artist 1", Songs: []*Song{song1}}
	album2 := &Album{ID: "2", Title: "Album 2", Artist: "Artist 2", Songs: []*Song{song2}}
	album3 := &Album{ID: "3", Title: "Album 3", Artist: "Artist 3", Songs: []*Song{song3}}

	// Create artists
	artist1 := &Artist{ID: "1", Name: "Artist 1", Albums: []*Album{album1}}
	artist2 := &Artist{ID: "2", Name: "Artist 2", Albums: []*Album{album2}}
	artist3 := &Artist{ID: "3", Name: "Artist 3", Albums: []*Album{album3}}

	// Add artists to library
	service.MusicLibrary.AddArtist(artist1)
	service.MusicLibrary.AddArtist(artist2)
	service.MusicLibrary.AddArtist(artist3)

	// Register users
	service.UserManager.RegisterUser(user1)
	service.UserManager.RegisterUser(user2)

	// Login user
	loggedInUser := service.UserManager.LoginUser("john@example.com", "password123")
	if loggedInUser != nil {
		fmt.Printf("User logged in: %s\n", loggedInUser.Username)
	} else {
		fmt.Println("Invalid username or password.")
	}

	// Search songs
	searchResults := service.MusicLibrary.SearchSongs("Song")
	fmt.Println("\nSearch Results:")
	for _, song := range searchResults {
		fmt.Printf("Song: %s - %s\n", song.Title, song.Artist)
	}

	// Create playlist
	playlist := NewPlaylist("1", "My Playlist", loggedInUser)
	playlist.AddSong(song1)
	playlist.AddSong(song2)
	loggedInUser.AddPlaylist(playlist)

	// Get recommendations
	recommendations := service.MusicRecommender.RecommendSongs(loggedInUser)
	fmt.Println("\nRecommended Songs:")
	for _, song := range recommendations {
		fmt.Printf("Song: %s - %s\n", song.Title, song.Artist)
	}

	// Create and use music player
	player := NewMusicPlayer()
	player.PlaySong(song1)
	player.PauseSong()
	player.SeekTo(60)

	// Display user playlists
	fmt.Println("\nUser Playlists:")
	for _, p := range loggedInUser.Playlists {
		fmt.Printf("Playlist: %s\n", p.Name)
		fmt.Println("Songs:")
		for _, s := range p.Songs {
			fmt.Printf("- %s\n", s.Title)
		}
	}
}

File: solutions/java/src/musicstreamingservice/MusicStreamingServiceDemo.java
package musicstreamingservice;

import java.time.Duration;
import java.util.List;

public class MusicStreamingServiceDemo {
    public static void run() {
        MusicStreamingService musicStreamingService = new MusicStreamingService();

        // Create users
        User user1 = musicStreamingService.registerUser( "john@example.com", "password123");
        User user2 = musicStreamingService.registerUser( "jane@example.com", "password456");

        // Create artists
        Artist artist1 = musicStreamingService.addNewArtist("Artist 1");
        Artist artist2 = musicStreamingService.addNewArtist("Artist 2");

        // Create albums
        Album album1 = musicStreamingService.addNewAlbum("Album 1", artist1.getId());
        Album album2 = musicStreamingService.addNewAlbum("Album 2", artist2.getId());

        // Create songs
        Song song1 = musicStreamingService.addNewSong("Song 1", artist1.getId(), album1.getId(), Duration.ofMinutes(4));
        Song song2 = musicStreamingService.addNewSong("Song 2", artist2.getId(), album2.getId(), Duration.ofMinutes(3));
        Song song3 = musicStreamingService.addNewSong("Song 3", artist2.getId(), album2.getId(), Duration.ofMinutes(5));

        // Add albums to artists
        musicStreamingService.addAlbumToArtist(artist1.getId(), album1.getId());
        musicStreamingService.addAlbumToArtist(artist2.getId(), album2.getId());

        // Add songs to albums
        musicStreamingService.addSongsToAlbum(album1.getId(), List.of(song1.getId(), song2.getId()));
        musicStreamingService.addSongsToAlbum(album2.getId(), List.of(song3.getId()));

        // Create playlists
        Playlist playlist1 = musicStreamingService.createPlaylist(user1.getId(), "Playlist 1");
        Playlist playlist2 = musicStreamingService.createPlaylist(user1.getId(), "Playlist 2");

        // Add songs to playlist
        musicStreamingService.addSongToPlaylist(user1.getId(), playlist1.getId(), song1.getId());
        musicStreamingService.addSongToPlaylist(user1.getId(), playlist1.getId(), song2.getId());

        // Search for songs
        List<Song> searchResults = musicStreamingService.searchSongs("Song");
        System.out.println("Search Results:");
        for (Song song : searchResults) {
            System.out.println("Song: " + song.getTitle() + " - " + song.getArtistName());
        }

        // Play a song
        musicStreamingService.playSong(user2.getId(), song1.getId());

        // Pause the song
        musicStreamingService.pauseSong(user2.getId());
    }
}
